src/replication.c

```diff
@@ -926,53 +1130,68 @@ accept_mgmt_conns(int epfd, int sfd) {
 	} \
 }
 
-replica_t *
-get_replica(int mgmt_fd, spec_t **s)
+/*
+ * For mgmt connection only..
+ * initial state is write_io_send_hdr, which will write header.
+ * it transitions to write_io_send_data based on length in hdr.
+ * once data is written, it will change io_state to READ_IO_RESP_HDR
+ * to read data from replica.
+ */
+static int
+write_io_data(replica_t *replica, io_event_t *wevent)
 {
-	replica_t *replica = NULL;
-	spec_t *spec = NULL;
-	*s = NULL;
-	MTX_LOCK(&specq_mtx);
-	TAILQ_FOREACH(spec, &spec_q, spec_next) {
-		MTX_LOCK(&spec->rq_mtx);
-		TAILQ_FOREACH(replica, &spec->rq, r_next) {
-			if(replica->mgmt_fd == mgmt_fd) {
-				*s = spec;
+	int fd = wevent->fd;
+	int *state = wevent->state;
+	zvol_io_hdr_t *write_hdr = wevent->io_hdr;
+	void **data = wevent->io_data;
+	int *write_count = wevent->byte_count;
+	uint64_t reqlen, count;
+	int errorno = 0, fd_closed = 0;
+	int donecount = 0;
+
+	switch(*state) {
+		case WRITE_IO_SEND_HDR:
+			reqlen = sizeof (zvol_io_hdr_t) - (*write_count);
+			count = write_data(fd, ((uint8_t *)write_hdr) + (*write_count), reqlen, &errorno, &fd_closed);
+			CHECK_AND_ADD_BREAK_IF_PARTIAL(fd_closed, (*write_count), count, reqlen);
+
+			*write_count = 0;
+			*state = WRITE_IO_SEND_DATA;
+			if (errorno == EAGAIN || errorno == EWOULDBLOCK)
 				break;
+		case WRITE_IO_SEND_DATA:
+			reqlen = write_hdr->len - (*write_count);
+			if (reqlen != 0) {
+				count = write_data(fd, ((uint8_t *)(*data)) + (*write_count), reqlen, &errorno, &fd_closed);
+				CHECK_AND_ADD_BREAK_IF_PARTIAL(fd_closed, (*write_count), count, reqlen);
 			}
-		}
-		if (replica != NULL) {
-			MTX_UNLOCK(&spec->rq_mtx);
-			break;
-		}
-		TAILQ_FOREACH(replica, &spec->rwaitq, r_waitnext) {
-			if(replica->mgmt_fd == mgmt_fd) {
-				*s = spec;
+			if (errorno == EAGAIN || errorno == EWOULDBLOCK)
```

Comment:
this EAGAIN can be removed in case of WRITE. We can move to READ state.
And, if EPOLLOUT comes, we need not do anything as the IO in the head of list is waiting on READ state.